### PR TITLE
Support limit offset as ExecNode

### PR DESCRIPTION
**Pull request by changhiskhan**

### CODE REVIEW

File: rust/src/io/exec/limit.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                    } else {
+                        // if this batch is longer than remaining limit
+                        // then slice up to the remaining limit
+                        b = b.slice(0, lim as usize);
+                        lim = 0;
+                    }
+                }
+
+                if tx.is_closed() {
+                    eprintln!("ExecNode(Take): channel closed");
```
Comment (eddyxu) [inline rust/src/io/exec/limit.rs]: Change error message to reflect the Node type?

File: rust/src/io/exec/limit.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    ) -> Self {
+        let (tx, rx) = mpsc::channel(4);
+        let limit = limit.unwrap_or(0).clone();
+        let offset = offset.unwrap_or(0).clone();
+        let io_thread = tokio::spawn(async move {
+            child.try_fold((offset, limit, tx), |(mut off, mut lim, tx), mut b: RecordBatch | async move {
+                let nrows = b.num_rows() as i64;
+                if off > 0 {
+                    if off > nrows {
+                        // skip this batch if offset is more than num rows
```
Comment (eddyxu) [inline rust/src/io/exec/limit.rs]: let's add a todo to skip the read from child entirely if `off > nrows`?

File: rust/src/io/exec/limit.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+
+impl Stream for Limit {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::into_inner(self).rx.poll_recv(cx)
+    }
+}
```
Comment (eddyxu) [inline rust/src/io/exec/limit.rs:112]: Add some tests?

File: rust/src/io/exec/limit.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+pub(crate) struct Limit {
+    rx: Receiver<Result<RecordBatch>>,
+    _io_thread: JoinHandle<()>,
+}
+
+impl Limit {
+    /// Create a new execution node to handle limit offset.
+    pub fn new(
+        child: impl ExecNode + Unpin + Send + 'static,
+        limit: Option<i64>,
```
Comment (eddyxu) [inline rust/src/io/exec/limit.rs:39]: should `limit` not a Option value

File: rust/src/io/exec/limit.rs (hunk 1 above)
Comment (changhiskhan) [reply rust/src/io/exec/limit.rs]: fixed

File: rust/src/io/exec/limit.rs (hunk 2 above)
Comment (changhiskhan) [reply rust/src/io/exec/limit.rs]: this is only useful without any predicates so probably won't be very impactful

File: rust/src/io/exec/limit.rs (hunk 4 above)
Comment (changhiskhan) [reply rust/src/io/exec/limit.rs:39]: offline discussion: postgres supports offset without limit

File: rust/src/dataset/scanner.rs [hunk 5]
```diff
@@ -147,6 +147,17 @@ impl<'a> Scanner {
                 Arc::new(projection.clone()),
                 flat_knn_node,
             ))
+        } else if self.limit.is_some() || self.offset.is_some() {
```
Comment (eddyxu) [inline rust/src/dataset/scanner.rs:150]: Could this work with other scanner, i.e., KNN?
for example, should it be the front of the `ExecNode` so far.

**Review by eddyxu — APPROVED** [review]
🚢

**Pull request merged**